### PR TITLE
Added code

### PR DESCRIPTION
**Pull request by calnix**
Please take a look at the Fractional Wrapper contract. I'm not too sure if I have the correct idea. 
Thanks!

### CODE REVIEW

File: src/FractionalWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+@title Fractional Wrapper
+@author Calnix
+@dev Contract allows users to exchange a pre-specified ERC20 token for some other wrapped ERC20 tokens.
+@notice Wrapped tokens will be burned, when user withdraws their deposited tokens.
+*/
+
+contract FractionalizedWrapper is ERC20Mock {
+
+    ///@dev Exchange rate at inception: 1 DAI = 1 yvDAI
+    uint exRate = 1*10**27;
```
Comment (alcueca) [inline src/FractionalWrapper.sol]: Using `1e27` notation is simpler.

File: src/FractionalWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        asset = asset_;
+    }
+
+
+    /// @notice User to deposit underlying tokens for FractionalWrapper tokens
+    /// @dev Mints shares Vault shares to receiver by depositing exactly amount of underlying tokens
+    /// @param assets The amount of underlying tokens to deposit
+    /// @param receiver Address of receiver of Fractional Wrapper shares
+    function deposit(uint assets, address receiver) external returns(uint shares) {       
+        shares = convertToShares(assets);
```
Comment (alcueca) [inline src/FractionalWrapper.sol:47]: You calculate this twice.

File: src/FractionalWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        //transfer assets
+        bool success = asset.transfer(receiver, assets);
+        require(success, "Transfer failed!"); 
+
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+
+    ///@dev set exchange rate 
+    function setExchangeRate(uint exRate_) public {
```
Comment (alcueca) [inline src/FractionalWrapper.sol]: This function should be restricted to the contract owner.

File: src/FractionalWrapper.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        exRate = exRate_;
+    }
+
+    ///@notice calculate how much yvDAI user should get based on exchange rate
+    ///@dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    function convertToShares(uint assets) public view returns(uint shares){
+        return (assets)*(exRate / 10**27);
```
Comment (alcueca) [inline src/FractionalWrapper.sol]: This ordering of operators will zero all decimals in `exRate`

File: src/FractionalWrapper.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    ///@notice calculate how much DAI user should get based on exchange rate
+    ///@dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    function convertToAssets(uint shares) public view returns(uint assets){
+        return (shares)/(exRate / 10**27);
```
Comment (alcueca) [inline src/FractionalWrapper.sol]: The arithmetic is not right, it's not the reverse of `convertToShares`. Have another go.

**Review by alcueca — CHANGES_REQUESTED** [review]
You are in the right path. You should now revisit the arithmetic and the use of access control contracts such as Ownable.sol or AccessControl.sol.

File: src/FractionalWrapper.sol (hunk 4 above)
Comment (calnix) [reply src/FractionalWrapper.sol]: Right, it should be (assets * exRate) / 10**27, so the division is carried out last.
was thinking in pen and paper math terms :stuck_out_tongue_closed_eyes:

**Comment by calnix** [comment]
1. Fleshed out Wrapper contract in full 
2. Made changes as directed by Alberto
3. Added testing contract

File: Readme.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+5. User can liquidate and get back underlying DAI, by burning wDAI.
+---> dai_qty = wDAI/ex_rate
+
+## Contracts
+1. DAI token - underlying
+2. Fwrapper - "vault w/ ex_rate" 
+
+Both contracts are ERC20Mock, to issue tokens. 
+Fwrapper must confirm to ERC4626 specification.
```
Comment (alcueca) [inline Readme.md]: Note that if you intend to conform to a specification, [all methods must be implemented](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4626.md#methods).

File: src/FractionalWrapper.sol [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+
+    /// @notice User to deposit underlying tokens for FractionalWrapper tokens
+    /// @dev Mints shares Vault shares to receiver by depositing exactly amount of underlying tokens
+    /// @param assets The amount of underlying tokens to deposit
+    /// @param receiver Address of receiver of Fractional Wrapper shares
+    function deposit(uint assets, address receiver) external returns(uint shares) {       
+        receiver = msg.sender;
+        shares = convertToShares(assets);
```
Comment (alcueca) [inline src/FractionalWrapper.sol:47]: Find out how to enable gas tracing in foundry and check the result of this call.

We generally avoid calling public functions from the same contract for cost reasons. Instead, we have internal functions with the logic, which are called by all public functions. [Example](https://github.com/yieldprotocol/vault-v2/blob/master/contracts/oracles/chainlink/ChainlinkMultiOracle.sol).

File: Readme.md (hunk 6 above)
Comment (calnix) [reply Readme.md]: Oops, my bad. I had assumed that while you conformed to the specs, you only implemented what was necessary. Will correct this.

**Comment by calnix** [comment]
@devtooligan
**TLDR: How can one call an inherited external function during testing?**

I set `convertToAssets()` and `convertToShares()` as internal functions. To be able to access them in my testing contract i had `abstract contract StateZero` inherit `FractionalWrapper`.

This allows me to access both convert functions directly as part of my testing. The issue faced is specific to the `**StateRateChangesTest**` scenario. In this scenario the exchange rate is updated by calling `setExchangeRate` - which is an **external** function. 

I can update the exchange rate of the deployed wrapper contract fine.

However, I am unable to call the **inherited** `setExchangeRate` function by any of the following 
1. pranking as deployer,
2. mockCall, 
3. Using stdstore - to directly set exchange rate.

In all cases, the visibility is blocked, as the function is external. 
This is necessary so that when I subsequently call `convertToAssets` & `convertToShares`, they would use the updated exchange rate.

The following tests fail because `convertToAssets` & `convertToShares` are still using the initial exchange rate. 
1. testWithdrawRateChange()
2. testRedeemRateChange()
4. testCannotWithdrawOrRedeemSameAmount()

Any ideas?

I'm aware I could call `wrapper.previewDeposit`/`wrapper.previewMint` to handle the necessary conversion and have the tests pass; as they are external and accessible. 

But I'm thinking this is a problem that would likely rear up again repeatedly.

File: src/FractionalWrapper.sol (hunk 7 above)
Comment (calnix) [reply src/FractionalWrapper.sol:47]: Yup, changed both converts to internal functions, that does break the last leg of my testing though. Reached out to Richie for guidance.

**Comment by alcueca** [comment]
What you do in that case is to create a mock contract, `FractionalWrapperMock` that inherits from `FractionalWrapper` and has public functions that all that they do is to call the internal functions. Then you test on the mock.

**Comment by alcueca** [comment]
```
function convert(...) public view returns (...) {
    return _convert(...);
}
```

File: src/FractionalWrapper.sol [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    /// @param exRate_ New exchange rate
+    function setExchangeRate(uint256 exRate_) external onlyOwner {
+        exRate = exRate_;
+    }
+
+    /// @notice calculate how much yvDAI user should get based on exchange rate
+    /// @dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    /// @param assets Amount of underlying tokens (assets) to be converted to wrapped tokens (shares)
+    /// Note: Apply division at the end as it results in the removal of 'decimal 0's
+    function convertToShares(uint256 assets) internal view returns(uint256 shares){
```
Comment (alcueca) [inline src/FractionalWrapper.sol]: Usually internal function names are preceded by an underscore.
```suggestion
    function _convertToShares(uint256 assets) internal view returns(uint256 shares){
```

File: src/FractionalWrapper.sol [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        return (assets * exRate) / 1e27;
+    }
+
+    /// @notice calculate how much DAI user should get based on exchange rate
+    /// @dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    /// @param shares Amount of wrapped tokens (shares)to be converted to underlying tokens (assets) 
+    /// Note: Apply division at the end as it results in the removal of 'decimal 0's
+    function convertToAssets(uint256 shares) internal view returns(uint256 assets){
```
Comment (alcueca) [inline src/FractionalWrapper.sol]:
```suggestion
    function _convertToAssets(uint256 shares) internal view returns(uint256 assets){
```

**Comment by alcueca** [comment]
If you are conforming to ERC4626, you need a public `convert*` functions, so you should have those and the internal `_convert*` ones. And yes, it's a bit of boilerplate.

**Review by alcueca — APPROVED** [review]
Once you fix the outstanding comments, you can merge. Well done on your first ERC-compliant contract 👏👏👏

**Comment by calnix** [comment]
1. Thanks for your response to the testing of internal functions
2. ERC4626 Compliance: convert* made public
3. StateRateChangesTest scenario has been modified to accept public convert* functions; all tests in this scenario pass successfully.

**Comment by calnix** [comment]
Deployed and Verified:

DAIToken: https://kovan-optimistic.etherscan.io/address/0x5e0e8ad46a8a81c6c83afea7e4ec43b2fdc59dc0
FractionalWrapper: https://kovan-optimistic.etherscan.io/address/0xbe3db9c07f2369ad881f60a16fed0a8af7047ebd

**Comment by calnix** [comment]
https://github.com/yieldprotocol/mentorship2022/issues/4
